bitcoin_keys/src/ecdsa/serialized_public_key.rs

```diff
@@ -0,0 +1,318 @@
+    // Note: we don't need to store Format because it is stored in zeroth byte.
+    data: [u8; 65],
+}
+
+impl SerializedPublicKey {
+    /// Serializes given public key.
+    ///
+    /// This function is intentionally private and written here as opposed to being a method on
+    /// `Legacy`. This avoids the potential cost of monomorphisation. However we still allow the
+    /// compiler to inline this as it may remove branches and just call the appropriate function.
+    #[inline]
+    pub(super) fn new(key: secp256k1::PublicKey, format: KeyFormat) -> Self {
+        let data = match format {
+            KeyFormat::Uncompressed => {
+                let data = key.serialize_uncompressed();
+                debug_assert_eq!(data[0], 4);
+                data
+            },
+            KeyFormat::Compressed => {
+                let serialized = key.serialize();
+                debug_assert!(serialized[0] == 2 || serialized[0] == 3, "unexpected first byte {}, should've been 2 or 3", serialized[0]);
+                let mut data = [0u8; 65];
+                data[..33].copy_from_slice(&serialized);
+                data
+            },
+        };
+
+        SerializedPublicKey {
+            data,
+        }
+    }
+
+    /// Returns the length of the slice.
+    ///
+    /// The returned value will be either 33 or 65, depending on the format of the key this was
+    /// created from.
+    #[inline]
+    pub fn len(&self) -> usize {
+        (**self).len()
+    }
+
+    /// Creates an iterator of bytes.
+    #[inline]
+    pub fn iter(&self) -> core::slice::Iter<'_, u8> {
+        (**self).iter()
+    }
+
+    /// Returns the serialized bytes as a slice.
+    ///
+    /// The length of the returned slice will be either 33 or 65, depending on the format of the
+    /// kye this was created from.
+    #[inline]
+    pub fn as_slice(&self) -> &[u8] {
+        &**self
+    }
+    
+    /// Returns raw pointer pointing to the beginning of the serialized bytes.
+    ///
+    /// To maintain memory safety the memory behind the pointer MUST NOT be accessed after `self`
+    /// is dropped or moved or mutably borrowed. You also MUST NOT write to the memory behind the
+    /// pointer. The memory is only valid for up to `self.len()` bytes.
+    #[inline]
+    pub fn as_ptr(&self) -> *const u8 {
+        (**self).as_ptr()
+    }
+}
+
+impl core::ops::Deref for SerializedPublicKey {
+    type Target = [u8];
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        // This produces a beautiful, short, branch-free assembly :)
+        //
+        // If the key format is uncompressed, the zeroth byte is 4, 2 or 3 otherwise.
+        // Thus the single bit is sufficient to decide the length.
+        // Multiplying by 8 we get 32 or 0, as needed.
+
+        // Debug check our assumptions.
+        debug_assert!(self.data[0] == 4 || self.data[0] == 2 || self.data[0] == 3);
+
+        &self.data[..(33 + (usize::from(self.data[0] & 4) * 8))]
```

Comment:
Very beautiful solution indeed!

Small nit for readability purposes:
```suggestion
        &self.data[..(33 + (usize::from(self.data[0] & 0b100) * 8))]
```